api/src/main/java/com/inrupt/client/ProblemDetails.java

```diff
@@ -0,0 +1,132 @@
+ * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+package com.inrupt.client;
+
+import com.inrupt.client.spi.JsonService;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.net.URI;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * A data class representing a structured problem description sent by the server on error response.
+ *
+ * @see <a href="https://www.rfc-editor.org/rfc/rfc9457">RFC 9457 Problem Details for HTTP APIs</a>
+ */
+public class ProblemDetails {
+    public static final String MIME_TYPE = "application/problem+json";
+    public static final String DEFAULT_TYPE = "about:blank";
+    private final URI type;
+    private final String title;
+    private final String details;
+    private final int status;
+    private final URI instance;
+
+    public ProblemDetails(
+        final URI type,
+        final String title,
+        final String details,
+        final int status,
+        final URI instance
+    ) {
+        // The `type` is not mandatory in RFC9457, so we want to set
+        // a default value here even when deserializing from JSON.
+        if (type != null) {
+            this.type = type;
+        } else {
+            this.type = URI.create(DEFAULT_TYPE);
+        }
+        this.title = title;
+        this.details = details;
+        this.status = status;
+        this.instance = instance;
+    }
+
+    public URI getType() {
+        return this.type;
+    };
+
+    public String getTitle() {
+        return this.title;
+    };
+
+    public String getDetails() {
+        return this.details;
+    };
+
+    public int getStatus() {
+        return this.status;
+    };
+
+    public URI getInstance() {
+        return this.instance;
+    };
+
+    public static ProblemDetails fromErrorResponse(
+            final int statusCode,
+            final Headers headers,
+            final byte[] body,
+            final JsonService jsonService
+    ) {
+        if (jsonService == null
+                || (headers != null && !headers.allValues("Content-Type").contains(ProblemDetails.MIME_TYPE))) {
+            return new ProblemDetails(
+                null,
+                HttpStatus.StatusMessages.getStatusMessage(statusCode),
+                null,
+                statusCode,
+                null
+            );
+        }
+        try {
+            // ProblemDetails doesn't have a default constructor, and we can't use JSON mapping annotations because
+            // the JSON service is an abstraction over JSON-B and Jackson, so we deserialize the JSON object in a Map
+            // and build the ProblemDetails from the Map values.
```

Comment:
Would it not be possible to use typed JSON deserialization if this class had a default constructor and conventionally named property methods?

I would prefer that to this Map-based logic.

Reply:
Let me give this a try

Reply:
Done in https://github.com/inrupt/solid-client-java/pull/1160/commits/7a59d45e5e05d59ad48246a5c77f02d26183325f via adding a package-private mutable data class used to initialize the immutable one.